build.assets/tooling/cmd/release-notes/release_notes.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	_ "embed"
+	"fmt"
+	"html/template"
+	"io"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+//go:embed template/release-notes.md.tmpl
+var tmpl string
+
+type tmplInfo struct {
+	Version     string
+	Description string
+}
+
+var (
+	releaseNotesTemplate = template.Must(template.New("release notes").Parse(tmpl))
+)
+
+type releaseNotesGenerator struct {
+	// releaseVersion is the version for the release.
+	// This will be compared against the version present in the changelog.
+	releaseVersion string
+}
+
+func (r *releaseNotesGenerator) generateReleaseNotes(md io.Reader) (string, error) {
+	desc, err := r.parseMD(md)
+	if err != nil {
+		return "", err
+	}
+
+	info := tmplInfo{
+		Version:     r.releaseVersion,
+		Description: desc,
+	}
+	var buff bytes.Buffer
+	if err := releaseNotesTemplate.Execute(&buff, info); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return buff.String(), nil
+}
+
+// parseMD is a simple implementation of a parser to extract the description from a changelog.
+// Will scan for the first double header and pull the version from that.
+// Will pull all information between the first and second double header for the description.
+func (r *releaseNotesGenerator) parseMD(md io.Reader) (string, error) {
```

Comment:
Do we have to implement markdown parsing "by hand"? Is there no library we can use?

It's fine either way as there's one less dependency but if we can do it with a couple of lines by using an existing library it'd be worth it IMO.

Reply:
There is a library that could be used: https://pkg.go.dev/rsc.io/markdown - this is the basis of what is used for parsing Go doc comments now that they have added more markdown syntax. But it seems overkill for the basic parsing we're doing here. I think it will end up larger and more complicated because you still need to iterate the structure to get the bits between the headings, but now there's a bunch of new types to deal with. But it might be worth a try to see what it looks like - I might be wrong.

Reply:
I did take a look at some libraries but they were all meant for rendering and extensibility of Markdown and not so much about searching. So instead of walking lines in a file it would be walking AST nodes. It might be less lines but I thought it might be more complicated.

I can give it a try with one or two to see if it's easier to use.